File: fsharp/bob/Bob.fs
module Bob

open System
open System.Text.RegularExpressions

let areAllWordsAcronyms words =
    let acronyms = ["OK"]
    Seq.forall (fun word -> List.contains word acronyms) words

let mapRegexMatchesToSeq groupCollection =
    groupCollection |> Seq.cast<Group> |> Seq.map (fun x -> x.Value)

let isAnyWordUpperCaseAndNotAcronym s =
    let m = Regex.Match(s, "[A-Z]{2,}") in
        if m.Success then
            let upperCaseWords = mapRegexMatchesToSeq m.Groups
            not (upperCaseWords |> areAllWordsAcronyms)
        else false


let (|IsYelling|_|) s =
    if isAnyWordUpperCaseAndNotAcronym s then Some IsYelling else None

let (|EndsWithQuestionMark|_|) (s:string) =
    if s.EndsWith "?" then Some EndsWithQuestionMark else None

let (|IsMeaningless|_|) (s:string) = 
    if String.IsNullOrWhiteSpace(s) then Some IsMeaningless else None

let hey statement =
    match statement with
    | IsYelling -> "Whoa, chill out!"
    | EndsWithQuestionMark -> "Sure."
    | IsMeaningless -> "Fine. Be that way!"
    | _ -> "Whatever."

File: csharp/bob/Bob.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;

public static class Bob
{
    enum AnswerTo
    {
        [Display(Name = "Whoa, chill out!")]
        Yelling,

        [Display(Name = "Sure.")]
        Question,

        [Display(Name = "Fine. Be that way!")]
        MeaninglessInput,

        [Display(Name = "Whatever.")]
        EverythingElse
    }

    public static string GetDisplayName(this Enum enumValue)
    {
        return enumValue.GetType().GetMember(enumValue.ToString())
                       .First()
                       .GetCustomAttribute<DisplayAttribute>()
                       .Name;
    }

    public static string Response(string statement)
    {
        statement = statement.Trim();

        if (IsYelling(statement))
        {
            return AnswerTo.Yelling.GetDisplayName();
        }

        if (IsQuestion(statement))
        {
            return  AnswerTo.Question.GetDisplayName();
        }

        if (IsMeaningless(statement))
        {
            return AnswerTo.MeaninglessInput.GetDisplayName();
        }
        
        return AnswerTo.EverythingElse.GetDisplayName();
    }


    private static bool IsQuestion(string s)
    {
        return s.EndsWith("?");
    }

    private static bool IsYelling(string s)
    {
        var acronyms = new List<string>(){
            "OK"
        };
        
        var regex = new Regex("[A-Z]{2,}");
        var match = regex.Match(s);

        var upperCaseWords = new List<string>();
        if (match.Success) 
        {
            foreach (Group upperCaseWordGroup in match.Groups)
            {
                upperCaseWords.Add(upperCaseWordGroup.Value);
            }
            
            var allWordsAreAcronyms = upperCaseWords.All(acronyms.Contains);
            return !allWordsAreAcronyms;
        }

        return false;
    }

    private static bool IsMeaningless(string s)
    {
        return string.IsNullOrWhiteSpace(s);
    }
}